### PR TITLE
fix to List.CartesianProduct node

### PR DESCRIPTION
**Pull request by aparajit-pratap**
### Purpose

Fixes regression caused by #9388 in `List.CartesianProduct` and adds missing test.

As a result of #9388 `for` loop behavior had changed. As a result of making `Get.ValueAtIndex` disallow array promotion of its list argument, for loops inside imperative language blocks no longer iterated over a single value. Previously the single value was getting promoted to an array before the `for` loop could iterate over it. This is because the for loop expression is first compiled to AST's as shown in this example: 
```
                x = 0;
                a = [10, 20, 30, 40]
                for(val in a)
                {
                    x = x + val;
                }

                Compiles down to:

                x = 0;
                a = [10, 20, 30, 40];
                val = null;
                %forloop_key = a.key;
                %forloop_expr = a;

                while( %forloop_key != null)
                {
                    val = Get.ValueAtIndex(%forloop_expr, %forloop_key); => Array promotion!!!
                    %forloop_key = %forloop_key + 1;
                    x = x + val;
                }
```
The fix is to add yet another ValueAtIndex like ZT method specifically for `for` loop compilation that does not have the `AllowArrayPromotion(false)` attribute.

### Declarations

Check these if you believe they are true

- [x] The code base is in a better state after this PR
- [ ] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [x] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [x] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions), and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.

### Reviewers

@mjkkirschner 
@QilongTang 

### FYIs

@smangarole

### CODE REVIEW

File: test/DynamoCoreTests/Nodes/ListTests.cs [hunk 1]
```diff
@@ -945,7 +945,34 @@ public void LaceShortest_Simple()
 
 		}
 
-		[Test]
+        [Test]
+        public void LaceCartesian_Simple()
```
Comment (mjkkirschner) [inline test/DynamoCoreTests/Nodes/ListTests.cs]: @aparajit-pratap can we also add a ds language level imperative test to assert the for loop behavior as well?

File: src/Libraries/DesignScriptBuiltin/Builtin.cs [hunk 2]
```diff
@@ -53,6 +53,11 @@ public static object ValueAtIndexInForLoop(IList list, int index)
                 return ValueAtIndex(list, index);
             }
 
+            public static object ValueAtIndexInForLoop(string stringList, int index)
```
Comment (mjkkirschner) [inline src/Libraries/DesignScriptBuiltin/Builtin.cs]: what code triggers this?

Comment (aparajit-pratap) [reply src/Libraries/DesignScriptBuiltin/Builtin.cs]: A `for` loop is now compiled to an AST containing this method. In other words this method is triggered when we run `for` loops.

Comment (aparajit-pratap) [reply src/Libraries/DesignScriptBuiltin/Builtin.cs]: More specifically it is triggered when we iterate over strings such as here:
```
r = [Imperative]
{
    s = "foo";
    for (x in "bar")
    {
         s  = s + x;
    }
    return = s;
};
```
Expected result: "foobar"

**Comment by aparajit-pratap** [comment]
@mjkkirschner I'm not happy with this solution of adding more `ValueAtIndex` methods just for the `for` loop but this is the simplest solution I have so far. Are you ok with it and the new method names or can you come up with better method names?

**Pull request merged**